traversal/scanner/GraphIterator.java

```diff
@@ -377,18 +378,72 @@ private void reset() {
         }
 
         private Forwardable<Vertex<?, ?>, Order.Asc> createIteratorFromEdges() {
-            List<Forwardable<Vertex<?, ?>, Order.Asc>> iters = new ArrayList<>();
-            edgeInputs.forEach((procedureEdge, vertex) -> iters.add(branch(vertex, procedureEdge)));
-            if (iters.size() == 1) return iters.get(0);
-            else return intersect(iterate(iters), ASC);
+            Map<Scopes.Scoped, List<ProcedureEdge.Native.Thing.RolePlayer>> rpEdgesByScope = new HashMap<>();
+            List<ProcedureEdge<?, ?>> otherEdges = new ArrayList<>();
+            edgeInputs.forEach((edge, vertex) -> {
+                if (edge.isRolePlayer()) {
+                    Scopes.Scoped scoped = scopes.getOrInitialise(edge.asRolePlayer().scope());
+                    rpEdgesByScope.computeIfAbsent(scoped, s -> new ArrayList<>()).add(edge.asRolePlayer());
+                } else otherEdges.add(edge);
+            });
+            return intersectEdges(rpEdgesByScope, otherEdges);
```

Comment:
we are doing the edge iterator intersection in two layers:
first we intersect RolePlayer edges using pairs of `(relation, role)`, which we guarantee to not occuply the same role across parallel edges. After this intersection we filter down to just the relations.

Then we intersect the remaining edges with the intersection of the role player edges.